corehq/ex-submodules/couchexport/export.py

```diff
@@ -313,19 +329,30 @@ def get_data(self):
     @classmethod
     def wrap_all_rows(cls, tables):
         """
-        Take a list of tuples (name, SINGLE_ROW) or (name, (ROW, ROW, ...))
+        Take a list of tuples (name, SINGLE_ROW) or (name, (ROW, ROW, ...)). The rows can be generators.
         """
-        ret = []
-        for name, rows in tables:
-            rows = list(rows)
-            if rows and (not hasattr(rows[0], '__iter__') or isinstance(rows[0], six.string_types)):
-                soft_assert_type_text(rows[0])
+
+        def _wrap_row(first_entry, rows):
+            if first_entry and (not hasattr(first_entry, '__iter__') or isinstance(first_entry, six.string_types)):
+                soft_assert_type_text(first_entry)
                 # `rows` is actually just a single row, so wrap it
                 rows = [rows]
-            ret.append(
-                (name, [cls(row) for row in rows])
-            )
-        return ret
+            return rows
+
+        wrapped_rows = []
+        for name, rows in tables:
+            if isinstance(rows, itertools.chain):
```

Comment:
similarly here I think it would be simpler to implement if you coerced to a generator up front rather than having parallel implementations

Reply:
https://github.com/dimagi/commcare-hq/pull/23782/commits/93eccfad3abfe1b80ce79d9bd34e7e27b1f197f7